vllm/v1/worker/gpu_worker.py

```diff
@@ -210,6 +229,311 @@ def profile(self, is_start=True):
+    INITIALIZE = b'\x01'  # Initialize cache and begin worker execution
+
+
+@dataclass
+class WorkerInitResponseType:
+    """
+    Request types defined as hex byte strings, so it can be sent over sockets
+    without separate encoding step.
+    """
+    READY = b'\x00'
+    NUM_BLOCKS = b'\x01'
+    INITIALIZE_SUCCESS = b'\x02'
+
+
+@dataclass
+class WorkerExecRequest:
+    """A directive from the core process to its worker processes.
+    
+	Wraps SchedulerOutput with a message type to distinguish between
+	regular work assignments and termination orders."""
+
+    class Type(Enum):
+        WORK = auto()
+        TERMINATE = auto()
+
+    message_type: Type
+    payload: Optional[SchedulerOutput]
+
+
+@dataclass
+class WorkerProcHandle:
+    proc: BaseProcess
+    initialization_input_path: str
+    initialization_output_path: str
+    model_output_mq_handle: Optional[Handle]
+
+    def ensure_termination(self):
+        """Ensure that the process is terminated.
+
+        Designed for use in the executor's shutdown method. Assumes worker
+        processes have already received termination requests.
+        Waits for processing, then sends termination and kill signals if needed.
+        """
+
+        def wait_for_termination(timeout):
+            start_time = time.time()
+            while time.time() - start_time < timeout:
+                if not self.proc.is_alive():
+                    return True
+                time.sleep(0.1)
+            return False
+
+        # First wait for graceful shutdown
+        if wait_for_termination(0.5):
+            return
+
+        # Send SIGTERM if still running
+        if self.proc.is_alive():
+            self.proc.terminate()
+            if wait_for_termination(1):
+                return
+
+        # Send SIGKILL if still running
+        if self.proc.is_alive():
+            os.kill(self.proc.pid, signal.SIGKILL)
+            if not wait_for_termination(1):
+                raise RuntimeError("Failed to terminate worker process")
+
+    def determine_num_available_blocks(self) -> Tuple[int, int]:
+        with make_zmq_socket(self.initialization_output_path,
+                             zmq.constants.PUSH) as send_socket, \
+             make_zmq_socket(self.initialization_input_path,
+                             zmq.constants.PULL) as recv_socket:
+
+            # Send message to determine the number of blocks
+            send_socket.send_multipart(
+                (WorkerInitRequestType.DETERMINE_NUM_BLOCKS, ))
+
+            # Receive response
+            type_frame, data_frame = recv_socket.recv_multipart(copy=False)
+            response_type = type_frame.buffer
+            response_data = data_frame.buffer
+            if response_type != WorkerInitResponseType.NUM_BLOCKS:
+                raise ValueError(f"Unknown RequestType: {response_type}")
+            return pickle.loads(response_data)
+
+    def initialize(self, num_gpu_blocks: int) -> bool:
```

Comment:
It looks like we could remove some of the duplicated code by having something like:

`def make_remote_request(request_type: WorkerInitRequestType, msg: Optional[]): `

The above function could handle all of the socket management.

Reply:
I think this code is a little too immature for a DRY pass so I'll hold off on this and we can refactor in a future PR